client/src/reducer.ts

```diff
@@ -99,39 +99,48 @@ export default function (state = INITIAL_STATE, action: any) {
+function addBit(state = Immutable.Map<string, any>(), bit: Bit) {
+  return setBitState(state, bit.postId).insert(bit);
+}
 
-const upvote = (state = Map<string, any>(), bitId: string) => resetVote(state, bitId)
-  .setIn(['bits', bitId, 'userVote'], USER_UPVOTE);
+function setVote(state = Immutable.Map<string, any>(), bitId: string, vote: Vote) {
+  return setBitState(state, bitId).edit(bit => bit.userVote = vote);
+}
 
-const downvote = (state = Map<string, any>(), bitId: string) => resetVote(state, bitId)
-  .setIn(['bits', bitId, 'userVote'], USER_DOWNVOTE);
+const upvote = (state = Immutable.Map<string, any>(), bitId: string) => setVote(state, bitId, USER_UPVOTE);
 
-const resetVote = (state = Map<string, any>(), bitId: string) => state
-  .setIn(['bits', bitId, 'userVote'], USER_DEFAULT_VOTE);
+const downvote = (state = Immutable.Map<string, any>(), bitId: string) => setVote(state, bitId, USER_DOWNVOTE);
 
-const changeSort = (state = Map<string, any>(), sort: string) => {
+const resetVote = (state = Immutable.Map<string, any>(), bitId: string) => setVote(state, bitId, USER_DEFAULT_VOTE);
+
+const changeSort = (state = Immutable.Map<string, any>(), sort: string) => {
   switch (sort) {
     case SORT_SCORE:
       return state.set('bits',
-        state.get('bits', Map<string, any>()).sortBy(
-          (bit: Map<string, any>) => -1 * (bit.get('upvotes', 0) - bit.get('downvotes', 0) + bit.get('userVote', 0)),
+        (state.get('bits', Immutable.OrderedMap()) as Immutable.OrderedMap<string, Readonly<Bit>>).sortBy(
+          bit => -1 * (bit.upvotes - bit.downvotes + bit.userVote)
         ));
     case SORT_DATE:
       return state.set('bits',
-        state.get('bits', Map<string, any>()).sortBy(
-          (bit: Map<string, any>) => -1 * bit.get('dateCreated', 0),
+        (state.get('bits', Immutable.OrderedMap()) as Immutable.OrderedMap<string, Readonly<Bit>>).sortBy(
+          bit => -1 * bit.dateCreated
         ));
     default:
       return state;
   }
 };
 
-const setProfile = (state = Map(), profile: any) => state.set('profile', profile);
+const setProfile = (state = Immutable.Map(), profile: any) => state.set('profile', profile);
 
-const receiveComments = (state = Map(), bitId: string, newComments: Map<string, any>) => state
-  .mergeIn(['comments'], Map(newComments.map(comment => [comment.get('postId'), comment])))
-  .updateIn(['bits', bitId, 'comments'], Set(), comments => comments.union(newComments.map(comment => comment.get('postId'))))
-  .setIn(['bits', bitId, 'isRequestingComments'], false);
+function receiveComments(state = Immutable.Map<string, any>(), bitId: string, newComments: Immutable.Set<any>) {
+  let stateWithComments = state.mergeIn(['comments'], Immutable.Map(newComments.map(comment => [comment.get('postId'), comment])));
+  return setBitState(stateWithComments, bitId)
+    .edit(bit => {
```

Comment:
FWIW I think this is exactly the same syntax as ImmerJS, not sure we can use that with TS.

Reply:
Copycats the lot of them